README.md
## Obsidian Siteswap

This plugin renders the "[Siteswap](https://en.wikipedia.org/wiki/Siteswap)" juggling notation via the Juggling Lab GIF server. Full documentation for the GIF server is [here](https://jugglinglab.org/html/animinfo.html).

This plugin requires an internet connection.

Straight forward siteswaps can be visualized via:

<pre>
```siteswap
(4x,2)*
```
</pre>

![](<https://jugglinglab.org/anim?redirect=true;height=200;width=200;pattern=(4%2C2x)*>)

If you want to specify more details than just the pattern, use the following form:

<pre>
```siteswap
pattern: 3
hands: (-25)(2.5).(25)(-2.5).(-25)(0).
colors: mixed
```
</pre>

![](<https://jugglinglab.org/anim?redirect=true;pattern=3;height=200;width=200;hands=(-25)(2.5).(25)(-2.5).(-25)(0).;colors=mixed>)

Details are specified via YAML - each line contains the name of a key, followed by a colon and a space, followed by a value.

Some attributes can be defined globally via the plugin settings. If global settings conflict with the details specified for a given animation, the global attributes are overwritten.

## Per Animation Settings

-   pattern: siteswap pattern to animate, in generalized siteswap notation.
-   bps: number of beats per second in the pattern, a floating point value. Default is automatically calculated from the pattern.
-   dwell: number of beats a catch is made prior to the subsequent throw. Values are floating-point numbers between 0.0 and 2.0; default value is 1.3.
-   hands: one of 'mills', 'inside', 'outside', 'half' or hand movement as a sequence of spatial coordinates for catches and throws and points in-between. Details [here](https://jugglinglab.org/html/sspanel.html).
-   body: body movement as a sequence of angles and spatial coordinates. Details [here](https://jugglinglab.org/html/sspanel.html).
-   colors: determines the coloring of the props. Each color is defined either by name or by its red/green/blue components on a 0-255 scale. For example if this setting is equal to {red} or {255,0,0}, the animator will use red balls. If you define several colors in a list, they are assigned to the balls in a cyclical manner. For example, {255,0,0}{0,255,0} means that ball 1 is red, ball 2 is green, ball 3 is red, ball 4 is green, and so on. Recognized color names are: black, blue, cyan, gray, green, magenta, orange, pink, red, yellow. Using the value mixed (no braces) will apply a pre-defined mix of colors. Default is {red}.
-   propdiam: diameter of the props, in centimeters. Default is 10.0.
-   prop: prop type to use. Recognized prop names are ball, image, and ring; default is ball.
-   gravity: acceleration of gravity, in cm/sec^2. Default is 980.0 (Earth standard).
-   bouncefrac: fraction of a ball's energy retained after bouncing off the ground (how much of its dropped height does it return to?). Values are floating point numbers greater than 0.0; default is 0.9.
-   hss: hand siteswap pattern to apply to the hands. See the [documentation](https://jugglinglab.org/html/HandSiteswapFeature.pdf).
-   handspec: (for hss mode only) assignment of hands to jugglers; see [documentation](https://jugglinglab.org/html/HandSiteswapFeature.pdf) for format.
-   dwellmax: (for hss mode only) whether to automatically maximize dwell time. Default value is true.
-   hold: (for hss mode only) whether to hold throws that can be held. Default value is false.

## Global Settings

-   width: Width of the animation, in pixels.
-   height: Height of the animation, in pixels.
-   scale: Scaling factor for the generated GIF. 1.0 performs no scaling.
-   fps: Number of frames per second in the generated GIF.
-   slowdown: Defines an overall time slowdown factor (e.g., slowdown: 1.0 is actual speed, slowdown: 2.0 is half actual speed).
-   showground: Whether to display the ground ("true", "false" or "auto").
-   camangle: Camera angles in degrees, given as one or a pair of angles. Example: camangle: (0,90). The first angle describes rotation of the camera around the juggler, and the second angle is the elevation angle given as degrees from directly overhead (i.e., 90 puts the camera on the same level as the juggler). Default value depends on the pattern.
-   hidejugglers: List of one or more jugglers to hide (i.e., not render) during animation. Examples: hidejugglers=1 or hidejugglers=(1,3).
-   stereo: Whether to display the pattern as a cross-eyed stereogram.


CHANGELOG.md
# 0.1.8 (Sun Jan 23 2022)

#### ⚠️ Pushed to `main`

- Add support for hands shortforms ([@tdresser](https://github.com/tdresser))

#### Authors: 1

- Timothy Dresser ([@tdresser](https://github.com/tdresser))

---

# 0.1.7 (Sun Jan 23 2022)

#### ⚠️ Pushed to `main`

- Address review feedback ([@tdresser](https://github.com/tdresser))

#### Authors: 1

- Timothy Dresser ([@tdresser](https://github.com/tdresser))

---

# 0.1.6 (Sat Jan 08 2022)

#### ⚠️ Pushed to `main`

- Siteswap Test.md exercises behavior. ([@tdresser](https://github.com/tdresser))

#### Authors: 1

- Timothy Dresser ([@tdresser](https://github.com/tdresser))

---

# 0.1.5 (Sat Jan 08 2022)

#### ⚠️ Pushed to `main`

- Build tweaks, remove usage of innerHTML. ([@tdresser](https://github.com/tdresser))

#### Authors: 1

- Timothy Dresser ([@tdresser](https://github.com/tdresser))

---

# 0.1.4 (Sat Jan 08 2022)

#### ⚠️ Pushed to `main`

- Remove empty styles.css ([@tdresser](https://github.com/tdresser))

#### Authors: 1

- Timothy Dresser ([@tdresser](https://github.com/tdresser))

---

# 0.1.3 (Sat Jan 08 2022)

#### ⚠️ Pushed to `main`

- Add License ([@tdresser](https://github.com/tdresser))

#### Authors: 1

- Timothy Dresser ([@tdresser](https://github.com/tdresser))

---

# 0.1.2 (Sat Jan 08 2022)

#### ⚠️ Pushed to `main`

- Should be ready for release ([@tdresser](https://github.com/tdresser))
- More versioning fixes ([@tdresser](https://github.com/tdresser))
- Actually use semantic versioning ([@tdresser](https://github.com/tdresser))
- Clean up versioning ([@tdresser](https://github.com/tdresser))
- Use auto for releasing ([@tdresser](https://github.com/tdresser))
- Build will copy artifacts into vault if OBSIDIAN_VAULT env variable exists. ([@tdresser](https://github.com/tdresser))
- Draft of documentation, done settings. ([@tdresser](https://github.com/tdresser))
- Cleanup. ([@tdresser](https://github.com/tdresser))
- Need to get settings changes to update gifs. ([@tdresser](https://github.com/tdresser))
- Prettier ([@tdresser](https://github.com/tdresser))
- Need to refresh gifs when settings change. ([@tdresser](https://github.com/tdresser))
- Refactor settings, there's a bug in reading them though ([@tdresser](https://github.com/tdresser))
- Not sure what broke everything ([@tdresser](https://github.com/tdresser))
- Working on settings ([@tdresser](https://github.com/tdresser))
- Working, now need to do settings, etc. ([@tdresser](https://github.com/tdresser))
- Trivial examples working ([@tdresser](https://github.com/tdresser))
- Parsing yaml ([@tdresser](https://github.com/tdresser))
- Broke something, but the idea is there... ([@tdresser](https://github.com/tdresser))
- Initial commit ([@tdresser](https://github.com/tdresser))

#### Authors: 1

- Timothy Dresser ([@tdresser](https://github.com/tdresser))

---

# 0.1.2 (Sat Jan 08 2022)

#### ⚠️ Pushed to `main`

- More versioning fixes ([@tdresser](https://github.com/tdresser))
- Actually use semantic versioning ([@tdresser](https://github.com/tdresser))
- Clean up versioning ([@tdresser](https://github.com/tdresser))
- Use auto for releasing ([@tdresser](https://github.com/tdresser))
- Build will copy artifacts into vault if OBSIDIAN_VAULT env variable exists. ([@tdresser](https://github.com/tdresser))
- Draft of documentation, done settings. ([@tdresser](https://github.com/tdresser))
- Cleanup. ([@tdresser](https://github.com/tdresser))
- Need to get settings changes to update gifs. ([@tdresser](https://github.com/tdresser))
- Prettier ([@tdresser](https://github.com/tdresser))
- Need to refresh gifs when settings change. ([@tdresser](https://github.com/tdresser))
- Refactor settings, there's a bug in reading them though ([@tdresser](https://github.com/tdresser))
- Not sure what broke everything ([@tdresser](https://github.com/tdresser))
- Working on settings ([@tdresser](https://github.com/tdresser))
- Working, now need to do settings, etc. ([@tdresser](https://github.com/tdresser))
- Trivial examples working ([@tdresser](https://github.com/tdresser))
- Parsing yaml ([@tdresser](https://github.com/tdresser))
- Broke something, but the idea is there... ([@tdresser](https://github.com/tdresser))
- Initial commit ([@tdresser](https://github.com/tdresser))

#### Authors: 1

- Timothy Dresser ([@tdresser](https://github.com/tdresser))

---

# 0.1.1 (Sat Jan 08 2022)

#### ⚠️ Pushed to `main`

- More versioning fixes ([@tdresser](https://github.com/tdresser))
- Actually use semantic versioning ([@tdresser](https://github.com/tdresser))
- Clean up versioning ([@tdresser](https://github.com/tdresser))
- Use auto for releasing ([@tdresser](https://github.com/tdresser))
- Build will copy artifacts into vault if OBSIDIAN_VAULT env variable exists. ([@tdresser](https://github.com/tdresser))
- Draft of documentation, done settings. ([@tdresser](https://github.com/tdresser))
- Cleanup. ([@tdresser](https://github.com/tdresser))
- Need to get settings changes to update gifs. ([@tdresser](https://github.com/tdresser))
- Prettier ([@tdresser](https://github.com/tdresser))
- Need to refresh gifs when settings change. ([@tdresser](https://github.com/tdresser))
- Refactor settings, there's a bug in reading them though ([@tdresser](https://github.com/tdresser))
- Not sure what broke everything ([@tdresser](https://github.com/tdresser))
- Working on settings ([@tdresser](https://github.com/tdresser))
- Working, now need to do settings, etc. ([@tdresser](https://github.com/tdresser))
- Trivial examples working ([@tdresser](https://github.com/tdresser))
- Parsing yaml ([@tdresser](https://github.com/tdresser))
- Broke something, but the idea is there... ([@tdresser](https://github.com/tdresser))
- Initial commit ([@tdresser](https://github.com/tdresser))

#### Authors: 1

- Timothy Dresser ([@tdresser](https://github.com/tdresser))

---

# (Sat Jan 08 2022)

#### ⚠️ Pushed to `main`

- Actually use semantic versioning ([@tdresser](https://github.com/tdresser))

main.ts
import { Plugin, MarkdownPostProcessorContext, parseYaml } from "obsidian";
import {
	SiteswapSettingTab,
	SiteswapSettings,
	DEFAULT_SETTINGS,
} from "settings";

const defaultSettingsObject = Object.assign({}, DEFAULT_SETTINGS);

// See https://github.com/jkboyce/jugglinglab/blob/c0226400230714571de078893c46aef734856ebc/source/jugglinglab/notation/MHNNotationControl.java#L24
function expandBuiltInHandString(
	hands: string | undefined
): string | undefined {
	if (hands == undefined) {
		return undefined;
	}
	const BUILT_IN_HAND_STRINGS: Map<string, string> = new Map(
		Object.entries({
			inside: "(10)(32.5).",
			outside: "(32.5)(10).",
			half: "(32.5)(10).(10)(32.5).",
			mills: "(-25)(2.5).(25)(-2.5).(-25)(0).",
		})
	);
	return BUILT_IN_HAND_STRINGS.get(hands.toLowerCase()) || hands;
}

export class SiteswapPlugin extends Plugin {
	settings: SiteswapSettings;

	static postprocessor = (settings: SiteswapSettings) => {
		return (
			source: string,
			el: HTMLElement,
			ctx: MarkdownPostProcessorContext
		) => {
			let failure: string | null = null;

			let yaml = null;
			try {
				// Replacing ':' with ': ' is a bit of a hack, but we know a priori that keys and values will
				// never contain colons, so this is safe.
				yaml = parseYaml(source.replaceAll(":", ": "));
			} catch (e) {
				failure = e.message;
			}

			if (failure != null) {
				// Pass.
			} else if (typeof yaml == "object") {
				if (!("pattern" in yaml)) {
					failure =
						'Invalid siteswap: the "pattern" attribute is required.';
				}
			} else if (typeof yaml == "number" || typeof yaml == "string") {
				yaml = {
					pattern: "" + yaml,
				};
			} else {
				failure = "Invalid siteswap.";
			}

			if (failure !== null) {
				const message = document.createElement("p");
				message.textContent = failure;
				message.style.color = "var(--text-error)";
				el.appendChild(message);
				return;
			}

			const paramsObject = { redirect: true, ...settings, ...yaml };
			const displayWidth = paramsObject.width;

			paramsObject.width = paramsObject.width / paramsObject.scale;
			paramsObject.height = paramsObject.height / paramsObject.scale;

			paramsObject.hands = expandBuiltInHandString(paramsObject.hands);

			// Don't pass default params. We want to fetched cached animations as much as possible,
			// and passing default params means we'll miss commonly cached animations.
			for (const key in paramsObject) {
				// Width and height are a bit tricky as the default used here isn't the default
				// for the gif generator. Special case those defaults here.
				if (key == "width" && paramsObject[key] == 400) {
					delete paramsObject["width"];
				} else if (key == "height" && paramsObject[key] == 450) {
					delete paramsObject["height"];
				} else {
					if (defaultSettingsObject[key] == paramsObject[key]) {
						delete paramsObject[key];
					}
				}
			}

			delete paramsObject.scale;

			const params = Object.keys(paramsObject)
				.map(
					(key) =>
						encodeURIComponent(key) +
						"=" +
						encodeURIComponent(paramsObject[key])
				)
				.join(";");

			const img = document.createElement("img");
			img.src = "https://jugglinglab.org/anim?" + params;
			img.style.width = displayWidth + "px";
			el.appendChild(img);
		};
	};

	async onload() {
		await this.loadSettings();

		console.log("loading siteswap plugin");
		this.registerMarkdownCodeBlockProcessor(
			"siteswap",
			SiteswapPlugin.postprocessor(this.settings)
		);

		this.addSettingTab(new SiteswapSettingTab(this.app, this));
	}

	onunload() {
		console.log("unloading siteswap plugin");
	}

	async loadSettings() {
		this.settings = Object.assign(
			{},
			DEFAULT_SETTINGS,
			await this.loadData()
		);
	}

	async saveSettings() {
		await this.saveData(this.settings);
	}
}

export default SiteswapPlugin;


esbuild.config.mjs
import esbuild from "esbuild";
import process from "process";
import builtins from "builtin-modules";
import path from "path";
import fs from "fs/promises";
import console from "console";

const banner = `/*
THIS IS A GENERATED/BUNDLED FILE BY ESBUILD
if you want to view the source, please visit the github repository of this plugin
*/
`;

async function writeToPluginDirIfNeeded() {
	if (process.env.OBSIDIAN_VAULT) {
		const plugin_dir = path.join(
			process.env.OBSIDIAN_VAULT,
			".obsidian",
			"plugins",
			"obsidian-siteswap"
		);
		try {
			await fs.stat(plugin_dir);
		} catch {
			await fs.mkdir(plugin_dir);
		}

		const hotreload_path = path.join(plugin_dir, ".hotreload");

		try {
			await fs.stat(hotreload_path);
		} catch {
			await fs.writeFile(hotreload_path, "");
		}
		await fs.copyFile("main.js", path.join(plugin_dir, "main.js"));
		await fs.copyFile(
			"manifest.json",
			path.join(plugin_dir, "manifest.json")
		);
		console.log("Copied to vault.");
	}
}

const prod = process.argv[2] === "production";

esbuild
	.build({
		banner: {
			js: banner,
		},
		entryPoints: ["main.ts"],
		bundle: true,
		external: ["obsidian", "electron", ...builtins],
		format: "cjs",
		watch: prod
			? false
			: {
					onRebuild(error, result) {
						writeToPluginDirIfNeeded();
					},
			  },
		target: "es2016",
		logLevel: "info",
		sourcemap: prod ? false : "inline",
		treeShaking: true,
		outfile: "main.js",
	})
	.catch((e) => {
		console.log(e);
		process.exit(1);
	});


settings.ts
import { SiteswapPlugin } from "main";
import { App, PluginSettingTab, Setting } from "obsidian";

export interface SiteswapSettings {
	width: number;
	height: number;
	scale: number;
	fps: number;
	stereo: boolean;
	slowdown: number;
	camangle: string;
	showground: "auto" | "true" | "false";
	hidejugglers: string;
}

export const DEFAULT_SETTINGS: SiteswapSettings = {
	width: 200,
	height: 225,
	scale: 0.5,
	fps: 33,
	stereo: false,
	slowdown: 2,
	camangle: "",
	showground: "auto",
	hidejugglers: "",
};

function stripNonNumerals(x: string): string {
	return x.replaceAll(/\D/g, "");
}

export class SiteswapSettingTab extends PluginSettingTab {
	plugin: SiteswapPlugin;

	constructor(app: App, plugin: SiteswapPlugin) {
		super(app, plugin);
		this.plugin = plugin;
	}

	// TODO: maybe write a declarative wrapper for this? Or refactor somehow.
	display(): void {
		const { containerEl } = this;

		containerEl.empty();

		containerEl.appendChild(
			createFragment((frag) => {
				frag.createEl("h2", { text: "Siteswap Settings" });
				frag.createEl("p", "????", (p) => {
					p.appendText("Detailed documentation can be found at ");
					p.createEl("a", {
						text: "jugglinglab.org/html/animinfo.html",
						href: "https://jugglinglab.org/html/animinfo.html",
					});
					p.appendText(`.`);
				});
				frag.createEl("p", {
					text: `These settings apply to all generated siteswap 
					animations, but can be overridden by specifying these parameters in the pattern 
					configuration.`,
				});
				frag.createEl("p", {
					text: `Note that existing animations won't update automatically after updating settings. 
			Turning the plugin off and on again is the easiest way to update everything.`,
				});
			})
		);

		new Setting(containerEl)
			.setName("width")
			.setDesc("Width of the animation, in pixels.")
			.addText((widget) =>
				widget
					.setValue("" + this.plugin.settings.width)
					.onChange(async (value) => {
						value = stripNonNumerals(value);
						widget.setValue(value);
						this.plugin.settings.width = parseInt(value);
						await this.plugin.saveSettings();
					})
			);

		new Setting(containerEl)
			.setName("height")
			.setDesc("Height of the animation, in pixels.")
			.addText((widget) =>
				widget
					.setValue("" + this.plugin.settings.height)
					.onChange(async (value) => {
						value = stripNonNumerals(value);
						widget.setValue(value);
						this.plugin.settings.height = parseInt(value);
						await this.plugin.saveSettings();
					})
			);

		new Setting(containerEl)
			.setName("scale")
			.setDesc(
				"Scaling factor for the generated GIF. 1.0 performs no scaling."
			)
			.addText((widget) =>
				widget
					.setValue("" + this.plugin.settings.scale)
					.onChange(async (value) => {
						value = stripNonNumerals(value);
						widget.setValue(value);
						this.plugin.settings.scale = parseInt(value);
						await this.plugin.saveSettings();
					})
			);

		new Setting(containerEl)
			.setName("fps")
			.setDesc("Number of frames per second in the generated GIF.")
			.addText((widget) =>
				widget
					.setValue("" + this.plugin.settings.fps)
					.onChange(async (value) => {
						value = stripNonNumerals(value);
						widget.setValue(value);
						this.plugin.settings.fps = parseInt(value);
						await this.plugin.saveSettings();
					})
			);

		new Setting(containerEl)
			.setName("slowdown")
			.setDesc(
				"Defines an overall time slowdown factor (e.g., slowdown: 1.0 is actual speed, slowdown: 2.0 is half actual speed)."
			)
			.addText((widget) =>
				widget
					.setValue("" + this.plugin.settings.slowdown)
					.onChange(async (value) => {
						value = stripNonNumerals(value);
						widget.setValue(value);
						this.plugin.settings.slowdown = parseInt(value);
						await this.plugin.saveSettings();
					})
			);

		const showGroundOptions: Record<"auto" | "true" | "false", string> = {
			true: "true",
			false: "false",
			auto: "auto",
		};
		new Setting(containerEl)
			.setName("showground")
			.setDesc("Whether to display the ground.")
			.addDropdown((dropdown) =>
				dropdown
					.addOptions(showGroundOptions)
					.setValue(this.plugin.settings.showground)
					.onChange(async (value) => {
						this.plugin.settings.showground = value as
							| "auto"
							| "true"
							| "false";
						await this.plugin.saveSettings();
					})
			);

		new Setting(containerEl)
			.setName("camangle")
			.setDesc(
				"Camera angles in degrees, given as one or a pair of angles. Example: camangle: (0,90). The first angle describes rotation of the camera around the juggler, and the second angle is the elevation angle given as degrees from directly overhead (i.e., 90 puts the camera on the same level as the juggler). Default value depends on the pattern."
			)
			.addText((text) =>
				text
					.setPlaceholder("(0,90)")
					.setValue(this.plugin.settings.camangle)
					.onChange(async (value) => {
						this.plugin.settings.camangle = value;
						await this.plugin.saveSettings();
					})
			);

		new Setting(containerEl)
			.setName("hidejugglers")
			.setDesc(
				"List of one or more jugglers to hide (i.e., not render) during animation. Examples: hidejugglers=1 or hidejugglers=(1,3)."
			)
			.addText((text) =>
				text
					.setPlaceholder("(1,3)")
					.setValue(this.plugin.settings.hidejugglers)
					.onChange(async (value) => {
						this.plugin.settings.hidejugglers = value;
						await this.plugin.saveSettings();
					})
			);

		new Setting(containerEl)
			.setName("stereo")
			.setDesc(
				"Whether to display the pattern as a cross-eyed stereogram."
			)
			.addToggle((toggle) =>
				toggle
					.setValue(this.plugin.settings.stereo)
					.onChange(async (value) => {
						this.plugin.settings.stereo = value;
						await this.plugin.saveSettings();
					})
			);
	}
}
